File: src/services/ai.js
import { getSetting } from './db'

const DEFAULT_MODEL = 'Qwen/Qwen2.5-1.5B-Instruct'

// Get Hugging Face configuration from settings
async function getConfig() {
  const apiKey = await getSetting('hfApiKey')
  const model = await getSetting('hfModel') || DEFAULT_MODEL
  return { apiKey, model }
}

/**
 * Test connection to Hugging Face API
 */
export async function testConnection() {
  const { apiKey, model } = await getConfig()

  if (!apiKey) {
    return { success: false, message: 'No API key configured. Add your Hugging Face API key in Settings.' }
  }

  try {
    const response = await fetch('/api/ai', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ action: 'test', apiKey, model })
    })

    const data = await response.json()
    return data
  } catch (e) {
    return { success: false, message: `Connection failed: ${e.message}` }
  }
}

/**
 * Generate weekly schedule from tasks
 * @param {Array} tasks - Array of task objects
 * @param {Array} projects - Array of project objects (for context)
 * @returns {Array} Schedule items with task_id, date, time_slot
 */
export async function generateWeeklySchedule(tasks, projects) {
  const { apiKey, model } = await getConfig()

  // Filter to incomplete tasks only
  const incompleteTasks = tasks.filter(t => !t.completed)

  if (incompleteTasks.length === 0) {
    return []
  }

  // If no API key, use fallback
  if (!apiKey) {
    console.log('No API key, using fallback schedule')
    return generateFallbackSchedule(incompleteTasks)
  }

  try {
    const response = await fetch('/api/ai', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        action: 'schedule',
        apiKey,
        model,
        tasks: incompleteTasks.map(t => ({
          id: t.id,
          title: t.title,
          priority: t.priority,
          due_date: t.due_date
        }))
      })
    })

    const data = await response.json()

    if (data.error) {
      console.error('Schedule generation error:', data.error)
      return generateFallbackSchedule(incompleteTasks)
    }

    return data.schedule || []
  } catch (e) {
    console.error('Schedule generation failed:', e)
    return generateFallbackSchedule(incompleteTasks)
  }
}

/**
 * Get AI task suggestions based on current context
 * @param {Array} tasks - Array of task objects
 * @param {Array} projects - Array of project objects
 * @returns {Array} Array of suggestion objects
 */
export async function getTaskSuggestions(tasks, projects) {
  const { apiKey, model } = await getConfig()

  const incompleteTasks = tasks.filter(t => !t.completed)

  if (incompleteTasks.length === 0) {
    return []
  }

  // Determine time of day context
  const hour = new Date().getHours()
  let timeContext = 'morning'
  if (hour >= 12 && hour < 17) timeContext = 'afternoon'
  else if (hour >= 17) timeContext = 'evening'

  // If no API key, use fallback
  if (!apiKey) {
    return getFallbackSuggestions(incompleteTasks, timeContext)
  }

  try {
    const response = await fetch('/api/ai', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        action: 'suggest',
        apiKey,
        model,
        tasks: incompleteTasks.slice(0, 10).map(t => ({
          id: t.id,
          title: t.title,
          priority: t.priority,
          due_date: t.due_date
        })),
        parameters: { availableMinutes: 60 }
      })
    })

    const data = await response.json()

    if (data.error || !data.suggestion) {
      return getFallbackSuggestions(incompleteTasks, timeContext)
    }

    // Return as array for compatibility
    return [{
      title: data.suggestion.title,
      reason: data.suggestion.reason,
      task_id: data.suggestion.task_id,
      type: 'task',
      estimated_time: '30 mins'
    }]
  } catch (e) {
    console.error('Task suggestions failed:', e)
    return getFallbackSuggestions(incompleteTasks, timeContext)
  }
}

/**
 * Get suggestion for what to work on next
 * @param {number} availableMinutes - Minutes available to work
 * @param {Array} tasks - Array of task objects
 * @param {Array} projects - Array of project objects
 * @returns {Object} Suggestion object with task_id, title, reason
 */
export async function getNextTaskSuggestion(availableMinutes, tasks, projects) {
  const { apiKey, model } = await getConfig()

  const incompleteTasks = tasks.filter(t => !t.completed)

  if (incompleteTasks.length === 0) {
    return null
  }

  // If no API key, return highest priority task
  if (!apiKey) {
    const top = getTopPriorityTask(incompleteTasks)
    return {
      task_id: top.id,
      title: top.title,
      reason: 'Highest priority task',
      can_complete: true
    }
  }

  try {
    const response = await fetch('/api/ai', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        action: 'suggest',
        apiKey,
        model,
        tasks: incompleteTasks.slice(0, 10).map(t => ({
          id: t.id,
          title: t.title,
          priority: t.priority,
          due_date: t.due_date
        })),
        parameters: { availableMinutes }
      })
    })

    const data = await response.json()

    if (data.error || !data.suggestion) {
      const top = getTopPriorityTask(incompleteTasks)
      return {
        task_id: top.id,
        title: top.title,
        reason: 'Highest priority task',
        can_complete: true
      }
    }

    return {
      ...data.suggestion,
      can_complete: true
    }
  } catch (e) {
    console.error('Next task suggestion failed:', e)
    const top = getTopPriorityTask(incompleteTasks)
    return {
      task_id: top.id,
      title: top.title,
      reason: 'Highest priority task',
      can_complete: true
    }
  }
}

// ============ Helper Functions ============

function getTopPriorityTask(tasks) {
  const priorityOrder = { Critical: 0, High: 1, Medium: 2, Low: 3 }
  return [...tasks].sort((a, b) => {
    const aPriority = priorityOrder[a.priority] ?? 4
    const bPriority = priorityOrder[b.priority] ?? 4
    return aPriority - bPriority
  })[0]
}

function getFallbackSuggestions(tasks, timeContext) {
  const priorityOrder = { Critical: 0, High: 1, Medium: 2, Low: 3 }

  const sortedTasks = [...tasks]
    .sort((a, b) => {
      const aPriority = priorityOrder[a.priority] ?? 4
      const bPriority = priorityOrder[b.priority] ?? 4
      return aPriority - bPriority
    })
    .slice(0, 3)

  return sortedTasks.map(task => ({
    title: task.title,
    reason: `${task.priority} priority task`,
    task_id: task.id,
    type: 'task',
    estimated_time: '30 mins'
  }))
}

function generateFallbackSchedule(tasks) {
  const priorityOrder = { Critical: 0, High: 1, Medium: 2, Low: 3 }
  const timeSlots = ['Morning (6am-12pm)', 'Afternoon (12pm-5pm)', 'Evening (5pm-9pm)']

  // Get next 7 days
  const dates = []
  const today = new Date()
  for (let i = 0; i < 7; i++) {
    const date = new Date(today)
    date.setDate(today.getDate() + i)
    dates.push(date.toISOString().split('T')[0])
  }

  const sortedTasks = [...tasks].sort((a, b) => {
    const aPriority = priorityOrder[a.priority] ?? 4
    const bPriority = priorityOrder[b.priority] ?? 4
    return aPriority - bPriority
  })

  const schedule = []
  let dateIndex = 0
  let slotIndex = 0

  for (const task of sortedTasks.slice(0, 14)) {
    schedule.push({
      task_id: task.id,
      date: dates[dateIndex],
      time_slot: timeSlots[slotIndex]
    })

    slotIndex++
    if (slotIndex >= timeSlots.length) {
      slotIndex = 0
      dateIndex = (dateIndex + 1) % dates.length
    }
  }

  return schedule
}

export default {
  testConnection,
  getTaskSuggestions,
  generateWeeklySchedule,
  getNextTaskSuggestion
}

File: api/ai.py
from http.server import BaseHTTPRequestHandler
import json
import requests

# HF Inference API endpoint
HF_API_URL = "https://api-inference.huggingface.co/models"

# Models available on free HF Inference API
DEFAULT_MODEL = "mistralai/Mistral-7B-Instruct-v0.3"

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_POST(self):
        content_length = int(self.headers.get('Content-Length', 0))
        post_data = self.rfile.read(content_length)

        try:
            data = json.loads(post_data.decode('utf-8'))
            action = data.get('action')
            api_key = data.get('apiKey')
            model = data.get('model', DEFAULT_MODEL)
            prompt = data.get('prompt', '')
            parameters = data.get('parameters', {})
            tasks = data.get('tasks', [])

            if not api_key:
                return self.send_json_response({'error': 'API key is required'}, 400)

            headers = {"Authorization": f"Bearer {api_key}"}

            if action == 'test':
                return self.handle_test(headers, model)
            elif action == 'generate':
                return self.handle_generate(headers, model, prompt, parameters)
            elif action == 'schedule':
                return self.handle_schedule(headers, model, tasks, parameters)
            elif action == 'suggest':
                return self.handle_suggest(headers, model, tasks, parameters)
            else:
                return self.send_json_response({'error': 'Invalid action'}, 400)

        except json.JSONDecodeError:
            return self.send_json_response({'error': 'Invalid JSON'}, 400)
        except Exception as e:
            return self.send_json_response({'error': str(e)}, 500)

    def query_hf(self, headers, model, payload):
        """Query the HF Inference API"""
        response = requests.post(
            f"{HF_API_URL}/{model}",
            headers=headers,
            json=payload,
            timeout=60
        )
        return response

    def handle_test(self, headers, model):
        """Test connection to HF Inference API"""
        try:
            payload = {
                "inputs": "Hello, respond with 'Connected!'",
                "parameters": {"max_new_tokens": 10}
            }

            response = self.query_hf(headers, model, payload)

            if response.status_code == 200:
                return self.send_json_response({
                    'success': True,
                    'message': f'Connected to HF Inference API! Model: {model}'
                })
            elif response.status_code == 401:
                return self.send_json_response({
                    'success': False,
                    'message': 'Invalid API key'
                })
            elif response.status_code == 503:
                # Model is loading
                return self.send_json_response({
                    'success': True,
                    'message': 'Connected! Model is loading, please wait ~20 seconds and try again.'
                })
            else:
                error_data = response.json() if response.text else {}
                error_msg = error_data.get('error', response.text)
                return self.send_json_response({
                    'success': False,
                    'message': f'API error ({response.status_code}): {error_msg}'
                })

        except requests.exceptions.Timeout:
            return self.send_json_response({
                'success': False,
                'message': 'Request timed out. Model may be loading.'
            })
        except Exception as e:
            return self.send_json_response({
                'success': False,
                'message': f'Error: {str(e)}'
            })

    def handle_generate(self, headers, model, prompt, parameters):
        """General text generation"""
        try:
            payload = {
                "inputs": prompt,
                "parameters": {
                    "max_new_tokens": parameters.get('maxTokens', 500),
                    "temperature": parameters.get('temperature', 0.7),
                    "return_full_text": False
                }
            }

            response = self.query_hf(headers, model, payload)

            if response.status_code == 200:
                result = response.json()
                if isinstance(result, list) and len(result) > 0:
                    text = result[0].get('generated_text', '')
                    return self.send_json_response([{'generated_text': text}])
                return self.send_json_response([{'generated_text': str(result)}])
            else:
                return self.send_json_response({'error': f'API error: {response.status_code}'}, 500)

        except Exception as e:
            return self.send_json_response({'error': str(e)}, 500)

    def handle_schedule(self, headers, model, tasks, parameters):
        """Generate weekly schedule from tasks"""
        if not tasks:
            return self.send_json_response({'schedule': []})

        # Get next 7 days
        from datetime import datetime, timedelta
        today = datetime.now()
        dates = [(today + timedelta(days=i)).strftime('%Y-%m-%d') for i in range(7)]

        # Build task list for prompt
        task_list = "\n".join([
            f"- ID:{t.get('id')}, Task: {t.get('title')}, Priority: {t.get('priority', 'Medium')}, Due: {t.get('due_date', 'None')}"
            for t in tasks[:12]
        ])

        prompt = f"""<s>[INST] You are a scheduling assistant. Create a weekly schedule for these tasks.

TASKS:
{task_list}

DATES: {', '.join(dates)}
TIME SLOTS: "Morning (6am-12pm)", "Afternoon (12pm-5pm)", "Evening (5pm-9pm)"

RULES:
- High priority tasks go earlier in the week
- Respect due dates (schedule before due date)
- Spread tasks across different days

Respond with ONLY a JSON array, no other text:
[{{"task_id": 1, "date": "YYYY-MM-DD", "time_slot": "Morning (6am-12pm)"}}]
[/INST]"""

        try:
            payload = {
                "inputs": prompt,
                "parameters": {
                    "max_new_tokens": 600,
                    "temperature": 0.3,
                    "return_full_text": False
                }
            }

            response = self.query_hf(headers, model, payload)

            if response.status_code == 200:
                result = response.json()
                text = ""
                if isinstance(result, list) and len(result) > 0:
                    text = result[0].get('generated_text', '')

                # Parse JSON from response
                schedule = self.extract_json_array(text)

                # Validate schedule items
                valid_schedule = []
                for item in schedule:
                    if item.get('task_id') and item.get('date') and item.get('time_slot'):
                        if item['date'] in dates:
                            valid_schedule.append(item)

                if valid_schedule:
                    return self.send_json_response({'schedule': valid_schedule})

            # Fallback if AI fails
            fallback = self.generate_fallback_schedule(tasks, dates)
            return self.send_json_response({'schedule': fallback, 'fallback': True})

        except Exception as e:
            fallback = self.generate_fallback_schedule(tasks, dates)
            return self.send_json_response({'schedule': fallback, 'fallback': True, 'error': str(e)})

    def handle_suggest(self, headers, model, tasks, parameters):
        """Suggest next task to work on"""
        if not tasks:
            return self.send_json_response({'suggestion': None})

        available_minutes = parameters.get('availableMinutes', 60)

        task_list = "\n".join([
            f"- ID:{t.get('id')}, Task: {t.get('title')}, Priority: {t.get('priority', 'Medium')}, Due: {t.get('due_date', 'None')}"
            for t in tasks[:10]
        ])

        prompt = f"""<s>[INST] You have {available_minutes} minutes. Which task should you work on?

TASKS:
{task_list}

Consider priority and due dates. Respond with ONLY this JSON format:
{{"task_id": <number>, "title": "<task name>", "reason": "<brief reason>"}}
[/INST]"""

        try:
            payload = {
                "inputs": prompt,
                "parameters": {
                    "max_new_tokens": 150,
                    "temperature": 0.3,
                    "return_full_text": False
                }
            }

            response = self.query_hf(headers, model, payload)

            if response.status_code == 200:
                result = response.json()
                text = ""
                if isinstance(result, list) and len(result) > 0:
                    text = result[0].get('generated_text', '')

                suggestion = self.extract_json_object(text)
                if suggestion:
                    return self.send_json_response({'suggestion': suggestion})

            # Fallback
            return self.fallback_suggestion(tasks)

        except Exception as e:
            return self.fallback_suggestion(tasks)

    def fallback_suggestion(self, tasks):
        """Return highest priority task as fallback"""
        if tasks:
            priority_order = {'Critical': 0, 'High': 1, 'Medium': 2, 'Low': 3}
            sorted_tasks = sorted(tasks, key=lambda t: priority_order.get(t.get('priority', 'Medium'), 2))
            top_task = sorted_tasks[0]
            return self.send_json_response({
                'suggestion': {
                    'task_id': top_task.get('id'),
                    'title': top_task.get('title'),
                    'reason': 'Highest priority task'
                },
                'fallback': True
            })
        return self.send_json_response({'suggestion': None})

    def extract_json_array(self, text):
        """Extract JSON array from text"""
        import re
        try:
            match = re.search(r'\[[\s\S]*?\]', text)
            if match:
                return json.loads(match.group())
        except:
            pass
        return []

    def extract_json_object(self, text):
        """Extract JSON object from text"""
        import re
        try:
            match = re.search(r'\{[\s\S]*?\}', text)
            if match:
                return json.loads(match.group())
        except:
            pass
        return None

    def generate_fallback_schedule(self, tasks, dates):
        """Generate simple schedule when AI fails"""
        priority_order = {'Critical': 0, 'High': 1, 'Medium': 2, 'Low': 3}
        sorted_tasks = sorted(tasks, key=lambda t: priority_order.get(t.get('priority', 'Medium'), 2))

        schedule = []
        time_slots = ['Morning (6am-12pm)', 'Afternoon (12pm-5pm)', 'Evening (5pm-9pm)']

        for i, task in enumerate(sorted_tasks[:14]):
            date_idx = i // 2 % len(dates)
            slot_idx = i % len(time_slots)
            schedule.append({
                'task_id': task.get('id'),
                'date': dates[date_idx],
                'time_slot': time_slots[slot_idx]
            })

        return schedule

    def send_json_response(self, data, status=200):
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode('utf-8'))
        return
